### PR TITLE
[WIP] Implement TypeORM service and associated Entities

### LINKED ISSUE #45: Implement Event Handlers
Thanks to #55, we now have the foundation for adding and updating data from the database. We now therefore need to implement the event handlers and create models as we go.

We also may need to add statistics (e.g. count up Kills on an alert for example) - therefore this marrys very well with #26.

# Handlers

## Priority
- [ ] ContinentLockEventHandler
- [ ] ContinentUnlockEventHandler
- [x] DeathEventHandler
- [x] FacilityControlEventHandler
- [x] MetagameEventEventHandler (naming lol)
- [ ] VehicleDestroyEventHandler (doesn't seem to exist! :O)

## Secondary
- [ ] PlayerFacilityCaptureEventHandler (named wrong in codebase)
- [ ] PlayerFacilityDefendEventHandler (ditto)
- [x] PlayerLoginEventHandler
- [x] PlayerLogoutEventHandler

## Meh

- [x] AchievementEarnedEventHandler
- [x] BattleRankUpEventHandler

# Models to create

## Alert
- [x] Alert
- [x] AlertControl
- [x] AlertDeath
- [x] AlertFacilityControl
- [ ] AlertPlayerFacility
- [x] AlertPopulation

## Aggregates
- [x] See #26 

## Misc
- [x] PlayerLogin
- [x] PlayerLogout

## Statics
Statics in my head is data that cannot change via the game, or requires customer service intervention to do so.
- [x] Player
- [x] Outfit

## Dynamics
Dynamics are values of data that can be changed at any time, e.g. an outfit's tag etc.
- [x] Player
- [x] Outfit

### LINKED ISSUE #48: Integrate Database connection
~So we've had some discussion and for now it seems logical to use TypeORM. It saves a lot of headaches creating schemas etc and it allows us to use entities which we just save.~

Changed to Mongoose

I've been assured that it doesn't have any massive performance implications, so for now I'm happy to use it. My personal experience with ORMs is that they're slow and comebersome, but that might be just a PHP land issue.

### PR DESCRIPTION
**Pull request by Maelstromeous**
Work towards the TypeORM service and associated Entities.

Closes #45 
Closes #48

### CODE REVIEW

**Comment by microwavekonijn** [comment]
I don't think the relations decorators like `@ManyToMany`, etc work for Mongo. A document database uses a different approach to relating data. For instance embedded documents that can be accessed by a property on the root document(also you can add multiple with arrays). The other is using the id field of a document.

Things like composite keys, joining, etc do not exist, everything is contained in their own document.

File: app/package.json [hunk 1]
```diff
@@ -39,6 +39,7 @@
     "reflect-metadata": "^0.1.13",
     "sequelize": "^5.21.7",
```
Comment (marci4) [inline app/package.json:40]: I think we can remove this?

File: app/src/entities/static/StaticClass.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
+import {Entity, Column} from 'typeorm';
+import {Faction} from '../../constants/faction';
+import {Loadout} from "../../constants/loadout";
+
+@Entity()
+export default class StaticClass {
+
+    // Maps to ingame loadout ID
+    @Column({
+        primary: true,
```
Comment (microwavekonijn) [inline app/src/entities/static/StaticClass.ts:10]: Not sure if this is correct. Mongo use `_id` for sharding and therefor it works a little bit different then primary keys in sql dbs

**Comment by NotThatBowser** [comment]
> I don't think the relations decorators like `@ManyToMany`, etc work for Mongo.

TypeORM's documentation gives the impression there's 1 way to work for all databases it works with (as there are no caveats) but it looks like document-based support is limited. Crucially the decorators don't add any magic to replicate how it would work on a relational database so the decorators won't work in this case.

Relevant issue: https://github.com/typeorm/typeorm/issues/655
Discussion of TypeORM with MongoDB: https://medium.com/@eliezer/typeorm-mongodb-review-8855903228b1

The above would suggest using the previous suggestion in developer-chat of using Mongoose.

For reference in how this would impact how models, queries are written:
Mongoose on Models: https://mongoosejs.com/docs/models.html
Mongoose on Subdocuments: https://mongoosejs.com/docs/subdocs.html

If the way of working with document-based DBs is relatively unfamiliar, and may require a significant change in the approach compared to the source project (meaning more time writing something new, rather than something better), it may be worth considering if changing back to using a relational DB could be simpler/faster.

**Comment by microwavekonijn** [comment]
The reason to use NoSQL is to make data collection simpler and with more options. Data modelling in SQL and document databases is very different as well as optimization, as there is a different approach to related data(i.e. embedded, no joins) and partitions(i.e. many collections).

**Comment by Maelstromeous** [comment]
> I don't think the relations decorators like `@ManyToMany`, etc work for Mongo. A document database uses a different approach to relating data. For instance embedded documents that can be accessed by a property on the root document(also you can add multiple with arrays). The other is using the id field of a document.
> 
> Things like composite keys, joining, etc do not exist, everything is contained in their own document.

I know at the very least foreign "keys" work, which is essentially just references to other entity IDs. But they're not keys in the normal sense.

I've yet to try the entities out, that's my next thing to do. You're probably very much correct in that the annotations don't make any difference. Hopefully it'll respect the enums though.

**Comment by Maelstromeous** [comment]
@NotThatBowser noted your comments, the benefit of writing this in orm format means if we wanted to move back to SQL we could. However it's been deemed that NoSQL is the better choice for this project for a variety of reasons, and while not everyone has experience in it, this is the best way to learn it for sure. There's going to be a few things we do wrong along the way, just need to adapt and learn from it.

File: app/src/utils/time.ts [hunk 3]
```diff
@@ -0,0 +1,3 @@
+export function getUnixTimestamp(): number {
+    return Math.round((new Date()).getTime() / 1000);
```
Comment (vostrnad) [inline app/src/utils/time.ts:2]:
```suggestion
    return Math.round(Date.now() / 1000);
```

**Comment by Maelstromeous** [comment]
Converting to draft, revisiting whether we should continue to use TypeORM or a combination of that and Mongoose for enforcing models.

**Comment by Maelstromeous** [comment]
PR has diverged greatly. Closing and re-opening.

**Pull request closed without merging**